li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -565,19 +570,79 @@ public void uncaughtException(Thread t, Throwable e) {
         _clusterGroup, (System.currentTimeMillis() - startTimeMs));
   }
 
+  @Override
+  synchronized public void close(long timeout, TimeUnit timeUnit) {
+    closeNoLock(timeout, timeUnit);
+  }
+
   @Override
   synchronized public void close(Duration timeout) {
-    close(timeout.toMillis(), TimeUnit.MILLISECONDS);
+    closeNoLock(timeout.toMillis(), TimeUnit.MILLISECONDS);
   }
 
   @Override
   synchronized public void wakeup() {
     throw new UnsupportedOperationException("Not implemented yet");
   }
 
-  synchronized public void reloadConfig(Map<String, Object> newConfigs) {
+  public LiKafkaConsumerConfig getCommonConsumerConfigs() {
+    return _commonConsumerConfigs;
+  }
+
+  public void reloadConfig(Map<String, String> newConfigs, UUID commandId) {
+    // Go over each consumer, close them, and update existing configs with newConfigs. Since each per-cluster consumer will be
+    // instantiated when the client began consuming from that cluster, we just need to clear the mappings and update the configs.
+    // since this method is invoked by marioClient, it should be non-blocking, so we create another thread doing above
+    LOG.info("Starting reloadConfig for LiKafkaConsumers in clusterGroup {} with new configs {}", _clusterGroup, newConfigs);
+    Thread t = new Thread(() -> {
+      closeExistingConsumers(newConfigs, commandId);
+    });
+    t.setDaemon(true);
+    t.setName("LiKafkaConsumer-reloadConfig-" + _clusterGroup.getName());
+    t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+      public void uncaughtException(Thread t, Throwable e) {
+        throw new KafkaException("Thread " + t.getName() + " throws exception", e);
+      }
+    });
+
+    t.start();
+  }
+
+  // Do synchronization on the method that will be closing the consumers since it will be handled by a different thread
+  synchronized void closeExistingConsumers(Map<String, String> newConfigs, UUID commandId) {
     ensureOpen();
```

Comment:
Do we want the client to throw an exception when Mario initiates reload configs on closed consumers? I wonder Mario should get an error response, instead?

Reply:
Agree, config reload is pushed from mario side, probably we should send an error message to mario in such cases and let mario decide what to do next. For now, I will just add a TODO here. Thanks for bringing this up.